### ISSUE TITLE
Naming conventions for async functions, or functions returning promises

### ISSUE THREAD

**Issue by erikarenhill** [dhis2/notes#27]
I sometimes find it hard to guess if a function is async/returning a promise or not. This is a call to arms to decide, or at least discuss pros/cons of suffixing functions (or other naming convention) with for example Async, e.g. `getFoobarAsync` instead of `getFoobar` if the function is async or returns a promise.

**Comment by erikarenhill** [comment]
Response from @Mohammer5 : what I don't like about async/await is that you normally have to add a try/catch block.. I prefer the then/catch way... But I guess that's just preference.

I think async functions should be grouped somewhere, like in an api.js or `utils/api` dir (so it's obvious that they're async).
IO/Side effects is one of the trickiest things to hanle in an app, so I think it's good to have a dedicated namespace for stuff like that, especially with redux apps where everything async happens in thunks/epics.
Has worked great for me so far

@erikarenhill : I think async/await makes code more readable than chaining/nesting then/etc
But either way, the function returning a promise can either be awaited or chained with then, so it might be a more proper question what a function returning a promise should be named

**Comment by erikarenhill** [comment]
Response from @Mohammer5 : nesting then
Shouldn't happen IMO.. Stuff like that normally can be simplified

> what a function returning a promise should be named
if it's an ajax function, I normally prefix with the method (get/post/put/delete)
For everything else I try to use verbs that indicate that time is involved, like `waitFor...`, `request...`, `fetch...`, `load...`

**Comment by erikarenhill** [comment]
A function named loadXXX could still be synchronously implemented even if it should not (remember back in the days using XMLHttpRequest you could force it to be synchronous).

In other languages such as C# there is a naming convention to suffix with Async even if it's a strongly typed language and the IDE will instantly tell you if a method returns a Task or Task<T> or not (which is awaitable the same way as a promise). My experience is that it makes life alot easier reading the code afterwards even if you could easily look up the type. I.E. it's not only helping you when you write the actual code, it also helps you later on getting back to the same code.

MDN also seems to use the Async suffix for some of their examples 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises

I've seen other suggestions on code styles to prefix with `when` to indicate that it's something that will resolve in the future, I like the idea but not sure I like the names that would come out of that.

**Comment by Mohammer5** [comment]
> A function named loadXXX could still be synchronously implemented even if it should not (remember back in the days using XMLHttpRequest you could force it to be synchronous).

This is true in theory, but I can't recall anything in the last few years where I have seen async code that should've been async.

> My experience is that it makes life alot easier reading the code afterwards even if you could easily look up the type.

I don't have the experience that's it's hard to ready async code because there are only three ways it should be done:

Way 1:
```js
const foo = await bar()
// do more stuff
```

Way 2
```js
await foo()
// do more stuff
```

Way 3
```js
foo()
    .then(/* do more stuff */)
```

I wouldn't mind trying it. I've read some people arguing about this and the only opinion that isn't just emotional is that people find it additional work to use the `Async` suffix, just like the hungarian notion kind of died in the FE world as well (you could argue the same about types).

My POV on this is:
1. Function names should indicate WHAT is happening and now HOW (I like names to be declarative).
2. Good function names hint that time is involved, so to me the real issue is good function naming and not missing an Async suffix

But I've never tried this. so I wouldn't mind trying to get my own experience with it. Let's see what the others will say.

**Comment by ismay** [comment]
I wouldn't mind having a suffix. I would personally prefer it though if we all want to do this to enforce it via eslint. Otherwise we might forget it, not use it, etc. and end up making things more confusing. There's an eslint issue on exactly this topic: https://github.com/eslint/eslint/issues/8531

So it seems like it's possible to enforce this via eslint. See here: https://github.com/eslint/eslint/issues/8531#issuecomment-302379564

**Comment by varl** [comment]
@Mohammer5 I think the problem is that it's hard to know if `bar` is asynchronous or not when you import it, e.g.:

```
import { bar } from 'dummy-dhis2-package'
```

It would be more clear if you did:

```
import { barAsync } from 'dummy-dhis2-package'
```

It reduces the mental overhead of keeping track of which imported functions are async, and removes the need to look it up (in code or docs) by explicitly stating it is async.

**Comment by Mohammer5** [comment]
@varl That's true, but `bar` is a very generic name. Is there any async function in our codebases that's unclear about being async or not?

We can try it though, if you guys want to go that way, I wouldn't mind too much.

**Comment by varl** [comment]
> @varl That's true, but `bar` is a very generic name. Is there any async function in our codebases that's unclear about being async or not?

Yeah `bar` is a contrived example. As for real examples, I know there's some functions in `d2` like `getInstance` and `getManifest` which are a bit ambiguous. Using `fetch` would imply asynchronicity but I don't think that `get` does.

> So it seems like it's possible to enforce this via eslint. See here: [eslint/eslint#8531 (comment)](https://github.com/eslint/eslint/issues/8531#issuecomment-302379564)

This is pretty cool though, and a lot easier to enforce than semantic function names.

**Comment by Mohammer5** [comment]
> Yeah bar is a contrived example. As for real examples, I know there's some functions in d2 like getInstance and getManifest which are a bit ambiguous. Using fetch would imply asynchronicity but I don't think that get does.

I see. We could argue about naming and implementation details here, but that's another discussion for another issue for another day.
Regardless of that, it is the way it is, so I totally get that the `Async` suffix would make it easier to reason about.

**Comment by varl** [comment]
To me it's important that it's enforceable by tools, after that I don't mind either way.

**Comment by varl** [comment]
After reading code which uses this pattern I realise that I do not like the pattern at all:

- Usage: https://github.com/dhis2/app-store/blob/c014c9f972d28437d8180014924f411480162d66/packages/server/src/routes/v1/apps/handlers/postCreateApp.js
- Definition: https://github.com/dhis2/app-store/blob/master/packages/server/src/data/addUserToOrganisationAsync.js

Main problems I have with it is that:

- It duplicates information:
    - `Async` in the filename: `addUserToOrganisationAsync.js`
    - `async` and `Async` in the function definition: `const addUserToOrganisationAsync = async (params, knex, transaction) => {}`
    - `Async` twice in the import line: `const createAppAsync = require('@data/createAppAsync')`
- It would make the most sense where it is more difficult to look up what the function returns (external libs) but we add it where lookup is easy (our own code).
- External libraries do not use the pattern so while the intent is to make our code cohesive, it does not deliver on that: https://github.com/dhis2/app-store/blob/c014c9f972d28437d8180014924f411480162d66/packages/server/src/routes/v1/apps/handlers/postCreateApp.js#L95
- Proper JSDoc gives enough hints to developer tools that a function returns a `Promise`, which is a clear indication that it is an async function.

**Comment by Mohammer5** [comment]
I share @varl 's opinion, here are my notes I've prepared on this topic for the Apr 11 meeting:

Stance: against it

* I always were able to get that from the context
* It's additional work when writing code
* It's not declarative
* It's very similary to hungrarian notation which is depricated
* If unclear what the function is returning, the naming or implementation can be improved
`getManifest` => `fetchManifest`
`getInstance` => `getInstanceSync` + `getInstanceAsync`
&nbsp;&nbsp;&nbsp;&nbsp;We start an application only after the d2 instance has been setup,
&nbsp;&nbsp;&nbsp;&nbsp;so this wouldn't even need to be async
* It will require a massive rewrite of many functions in the existing codebase (or create inconsistency)

**Comment by erikarenhill** [comment]
> async and Async in the function definition: const addUserToOrganisationAsync = async (params, knex, transaction) => {}

This could easily be written as 
`module.exports = async (params, knex, transaction) => {}` instead so that is really a non-issue imho. Since it's a default export that also removes the issues you are having with redundant information in the filenames, as it could be named addUserToOrganisation instead and then imported with the Async suffix

> Proper JSDoc gives enough hints to developer tools that a function returns a Promise, which is a clear indication that it is an async function.

^ Agree

As long as JSDoc are done properly I don't mind **not** using the async suffix.

Since node libraries nowadays by default use the sync suffix we should do the same instead of the other way around.


> All of the I/O methods in the Node.js standard library provide asynchronous versions, which are non-blocking, and accept callback functions. Some methods also have blocking counterparts, which have names that end with Sync.

^ https://nodejs.org/de/docs/guides/blocking-vs-non-blocking/

**Comment by varl** [comment]
> This could easily be written as
> `module.exports = async (params, knex, transaction) => {}` instead so that is really a non-issue imho. 

Sure, although that's not how it was written, so it introduces an implicit style rule where it's easy to make "mistakes".

I also _like_ having a name attached to a function and not directly assigning an unnamed function directly to the `modules.exports` as that can affect stack traces and debugging.

>Since node libraries nowadays by default use the sync suffix we should do the same instead of the other way around.

I think that's a solid argument.

**Comment by Mohammer5** [comment]
> > Since node libraries nowadays by default use the sync suffix we should do the same instead of the > other way around.
> 
> I think that's a solid argument.

I think it's a solid argument for node libraries! But not for libraries in general. In the node environment, most things can be done **BOTH** synchronously and asynchronously, with is certainly not the case for client libs! I'd be totally for this approach when writing server-side libraries, but that does not apply to FE libs

**Comment by erikarenhill** [comment]
> This could easily be written as
module.exports = async (params, knex, transaction) => {} instead so that is really a non-issue imho.

Sure, although that's not how it was written, so it introduces an implicit style rule where it's easy to make "mistakes".

@varl you are right that's not the way it's written ATM, but that's also because I continued to implement  the way it was started while waiting for the outcome from this note. Since it **could** be rewritten as the stated example I'd say we should discuss how we should do it instead how it has been done (which will be refactored either way)

**Comment by varl** [comment]
> @varl you are right that's not the way it's written ATM, but that's also because I continued to implement the way it was started while waiting for the outcome from this note.

@erikarenhill I did not intend to criticize, I apologize for being blunt if it came across as criticism.

This is the only place in our code base that I could find concrete examples of code to discuss. I think talking about abstract non-existing code is not very helpful and would prefer to talk about actual working code.

Since this code actually demonstrates how it would look and work in reality, it is a great case to study, play around with, and talk about.

>Since it **could** be rewritten as the stated example I'd say we should discuss how we should do it instead of using arguments to how it has been done in a currently bad way (which will be refactored either way)

I do not think it is bad currently. My point is that it _could_ be written in as many ways as there are developers, and rewritten in even more ways.

The most natural way to (and how most developers would) write the code I linked _is_ how it was written. Why would we choose to write it differently if this is the natural and clear way to write it?

I _want_ the code to be written as it is because it's the idiomatic way to write it. The suffix is _not_ idiomatic and it adds noise to otherwise clear code. Rewriting the code in a different style seems drastic when just dropping the suffix removes the problem.

Another option is to put the async code in a `async` folder and use e.g. `require('./handlers/async/addUserToOrganisation')` to avoid the `Async` suffix on functions _and_ on file names by just changing the structure instead of adding noise to the code.

**Comment by erikarenhill** [comment]
Please see this PR for how I actually want the code to look, then it would
make more sense to name the import as in previous examples (leaving the
filename until this note is resolved) :
https://github.com/dhis2/app-store/pull/22

I hope it makes sense

**Comment by varl** [comment]
> Please see this PR for how I actually want the code to look, then it would make more sense to name the import as in previous examples (leaving the filename until this note is resolved) : [dhis2/app-store#22](https://github.com/dhis2/app-store/pull/22) I hope it makes sense

It makes sense, but ultimately I think it is less clear and harder to read. Now the function name is in some cases 10 lines away from the function body.

I think the most clear is the variant you had in the first place. Here I chose to omit default exports because I find it does make the code base a lot easier to work with in terms of reading, debugging, and refactoring.

```
/**
 * Publish an app version to a channel
 * 
 * @function
 * @async
 *
 * @param {object} params The parameters used to publish an app version to a specific channel
 * @param {number} params.appVersionId The app version db id this media belongs to
 * @param {number} params.userId The id for the user which uploaded the media ("created by user id")
 * @param {number} params.imageType ImageType enum that determines if this is a logotype or image/screenshot
 * @param {string} params.fileName Original filename as when uploaded
 * @param {string} params.mime MIME type for the file, for example 'image/jpeg'
 * @param {object} knex DB instance of knex
 * @param {object} trx The transaction to operate on
 * @returns {Promise<AppVersionMediaResult>}
 */
exports.addAppVersionMedia = async ({...}) => {
    ...
}
```

There's an [`@async`](http://usejsdoc.org/tags-async.html) tag in JSDoc too, which would be useful to include.

**Comment by erikarenhill** [comment]
The `@async` tag was something I initially added to the JSDocs but removed before committing since the JSDoc because the JSDocs docs say you should not use it generally due to JSDoc automatically detect it.

> In general, you do not need to use this tag, because JSDoc automatically detects asynchronous functions and identifies them in the generated documentation. However, if you are writing a virtual comment for an asynchronous function that does not appear in your code, you can use this tag to tell JSDoc that the function is asynchronous.
http://usejsdoc.org/tags-async.html

**Comment by varl** [comment]
Are you writing JSDoc for the machine or for humans to read?

JSDoc is able to automatically detect that it is an async function, though it is still helpful to me as a human reading the doc string if it is there.

**Comment by erikarenhill** [comment]
> JSDoc is able to automatically detect that it is an async function, though it is still helpful to me as a human reading the doc string if it is there.

But as previously stated, you can read it in the function declaration and stating it again in the JSDocs makes it redundant? Lets move the JSDocs comments to the PR, belongs more there than here..

**Comment by Mohammer5** [comment]
@erikarenhill Interesting example. 

<hr />

I guess the param naming is a bit odd here, it's called `request`, I guess you mean `response`, right? Otherwise I would interpret that it should've been
```js
const isAuthenticated = (request) => request.then(response => response.auth.isAuthenticated)
```
But I'm certain you meant `response`

<hr />

To me, `isAuthenticated` should return a `Promise`. No idea how much contact you've had with functional reactive programming. To me it feels a lot more natural to make it like this:

```js
const isAuthenticated = reponse => new Promise(
  (resolve, reject) => response.auth.isAuthenticated ? resolve() : reject()
)

const checkEditAppsPermission = async () => {
  const app = await db('apps').where(something)
  return app.ownerId === whatever
}

const canUserEditApps = isAuthenticated(response)
  .then(checkEditAppsPermissions)
  .catch(notifyUser(`You lack some permissions`))
```

`isAuthenticated` is not async, but it behaves async, allowing for a nice control flow without having to clutter the code for checking whether the user is authenticated or not and you're able to split code into single responsibility units.

**Comment by erikarenhill** [comment]
> I guess the param naming is a bit odd here, it's called request, I guess you mean response, right? 

No it's request. Remember this specific example with regards to app store backend is when we receive the request from a browser/client with a JWT token that's being decoded. I see your point if we were talking about a FE app though :)

**Comment by varl** [comment]
**I think I vastly over-extrapolated the intention of this note**.

I assumed that the intent was to standardize on a naming convention with the `Suffix` across all our libraries and applications (front and back). This was why I vehemently argue about it.

If this is just the convention in a single app (app-store backend for now) I think that it is perfectly fine to do whatever the developers think is right. When I work with a code base I always try to follow the conventions established in it, and I think it's fine (and preferable) for applications of different sizes to have different conventions. Imagine if all 2k line apps looked like 20k line apps.

If it turns out to be wildly inappropriate it can always be refactored, with little discussion.

Where it's necessary for us to have consistent conventions across **all** of our applications it is in how we apply architectural paradigms. E.g. [Redux conventions](https://github.com/dhis2/notes/blob/master/decisions/2019/02/24-fe-redux-conventions.md), Forms, UI components, etc.

**Comment by Mohammer5** [comment]
> I assumed that the intent was to standardize on a naming convention with the Suffix across all our libraries and applications (front and back). This was why I vehemently argue about it.

I made the same assumption..

**Comment by varl** [comment]
We can drop this from the meeting and allow for naming convention variance between applications.
